File: modules/ctx-lite/tests/fixtures/signatures/sample.ts
/**
 * A comprehensive TypeScript module for signature extraction testing.
 */

export interface IProcessor {
  process(data: string): Promise<string>;
  validate(data: string): boolean;
  getName(): string;
}

export interface IConfig {
  name: string;
  version: string;
  debug: boolean;
}

/**
 * Application configuration class.
 */
export class Configuration implements IConfig {
  public name: string;
  public version: string;
  public debug: boolean;
  public timeout: number;

  constructor(name: string, version: string, debug: boolean = false) {
    this.name = name;
    this.version = version;
    this.debug = debug;
    this.timeout = 30000;
  }

  public validate(): boolean {
    return this.name.length > 0 && this.version.length > 0;
  }

  public setDebug(enabled: boolean): void {
    this.debug = enabled;
  }
}

/**
 * Base processor class for data handling.
 */
export abstract class BaseProcessor implements IProcessor {
  protected config: Configuration;
  protected cache: Map<string, string>;

  constructor(config: Configuration) {
    this.config = config;
    this.cache = new Map();
  }

  public async process(data: string): Promise<string> {
    const cached = this.cache.get(data);
    if (cached) return cached;

    const result = await this.doProcess(data);
    this.cache.set(data, result);
    return result;
  }

  protected abstract doProcess(data: string): Promise<string>;

  public abstract validate(data: string): boolean;

  public abstract getName(): string;

  public clearCache(): void {
    this.cache.clear();
  }
}

/**
 * Processor for text data.
 */
export class TextProcessor extends BaseProcessor {
  protected async doProcess(data: string): Promise<string> {
    return data.toUpperCase();
  }

  public validate(data: string): boolean {
    return typeof data === 'string' && data.length > 0;
  }

  public getName(): string {
    return 'TextProcessor';
  }
}

/**
 * Processor for JSON data.
 */
export class JsonProcessor extends BaseProcessor {
  protected async doProcess(data: string): Promise<string> {
    const parsed = JSON.parse(data);
    return JSON.stringify(parsed, null, 2);
  }

  public validate(data: string): boolean {
    try {
      JSON.parse(data);
      return true;
    } catch {
      return false;
    }
  }

  public getName(): string {
    return 'JsonProcessor';
  }

  public extractKeys(data: string): string[] {
    const parsed = JSON.parse(data);
    return Object.keys(parsed);
  }
}

/**
 * Factory for creating processors.
 */
export class ProcessorFactory {
  private static processors: Map<string, typeof BaseProcessor> = new Map([
    ['text', TextProcessor],
    ['json', JsonProcessor],
  ]);

  public static createProcessor(
    type: string,
    config: Configuration
  ): IProcessor {
    const ProcessorClass = this.processors.get(type);
    if (!ProcessorClass) {
      throw new Error(`Unknown processor type: ${type}`);
    }
    return new ProcessorClass(config);
  }

  public static registerProcessor(
    type: string,
    processorClass: typeof BaseProcessor
  ): void {
    this.processors.set(type, processorClass);
  }
}

/**
 * Main application manager.
 */
export class ApplicationManager {
  private config: Configuration;
  private processors: Map<string, IProcessor>;

  constructor(config: Configuration) {
    this.config = config;
    this.processors = new Map();
    this.initializeDefaultProcessors();
  }

  private initializeDefaultProcessors(): void {
    this.processors.set(
      'text',
      ProcessorFactory.createProcessor('text', this.config)
    );
    this.processors.set(
      'json',
      ProcessorFactory.createProcessor('json', this.config)
    );
  }

  public registerProcessor(name: string, processor: IProcessor): void {
    this.processors.set(name, processor);
  }

  public async processFile(
    filename: string,
    processorName: string
  ): Promise<string | null> {
    const processor = this.processors.get(processorName);
    if (!processor) {
      throw new Error(`Unknown processor: ${processorName}`);
    }

    try {
      const fs = require('fs').promises;
      const data = await fs.readFile(filename, 'utf-8');
      return await processor.process(data);
    } catch (error) {
      if (this.config.debug) {
        console.error(`Error processing ${filename}:`, error);
      }
      return null;
    }
  }

  public getProcessor(name: string): IProcessor | undefined {
    return this.processors.get(name);
  }

  public getConfig(): Configuration {
    return this.config;
  }
}

/**
 * Utility function for string encoding.
 */
export function encodeString(text: string): string {
  return Array.from(text)
    .map((c) => c.charCodeAt(0).toString(16))
    .join('-');
}

/**
 * Utility function for string decoding.
 */
export function decodeString(encoded: string): string {
  return encoded
    .split('-')
    .map((hex) => String.fromCharCode(parseInt(hex, 16)))
    .join('');
}

/**
 * Process stream data asynchronously.
 */
export async function processStream(
  streamHandler: AsyncIterable<any>
): Promise<{ processed: number; errors: number }> {
  const results = { processed: 0, errors: 0 };
  for await (const item of streamHandler) {
    try {
      results.processed++;
    } catch (error) {
      results.errors++;
    }
  }
  return results;
}

// Main execution
if (require.main === module) {
  const config = new Configuration('demo', '1.0', true);
  const manager = new ApplicationManager(config);
  console.log('TypeScript Application initialized');
}

File: modules/ctx-lite/tests/fixtures/signatures/sample.py
"""A comprehensive Python module for signature extraction testing."""

from typing import Optional, List, Dict, Any
from dataclasses import dataclass
from enum import Enum
import json


class ProcessorType(Enum):
    """Enumeration of available processor types."""
    TEXT = "text"
    JSON = "json"
    BINARY = "binary"


@dataclass
class Configuration:
    """Application configuration data class."""
    name: str
    version: str
    debug: bool = False
    timeout: int = 30


class DataProcessor:
    """Base class for data processing operations."""

    def __init__(self, config: Configuration):
        """Initialize processor with configuration."""
        self.config = config
        self.cache = {}

    def process(self, data: str) -> str:
        """Process input data and return result."""
        if data in self.cache:
            return self.cache[data]
        result = self._do_process(data)
        self.cache[data] = result
        return result

    def _do_process(self, data: str) -> str:
        """Internal processing method to be overridden."""
        raise NotImplementedError("Subclasses must implement _do_process")

    def clear_cache(self) -> None:
        """Clear the internal cache."""
        self.cache.clear()


class TextProcessor(DataProcessor):
    """Processor for text files."""

    def _do_process(self, data: str) -> str:
        """Convert text to uppercase."""
        return data.upper()

    def validate(self, data: str) -> bool:
        """Validate text data."""
        return len(data) > 0 and isinstance(data, str)


class JSONProcessor(DataProcessor):
    """Processor for JSON files."""

    def _do_process(self, data: str) -> str:
        """Parse and format JSON."""
        parsed = json.loads(data)
        return json.dumps(parsed, indent=2, sort_keys=True)

    def extract_keys(self, data: str) -> List[str]:
        """Extract all keys from JSON structure."""
        parsed = json.loads(data)
        return list(parsed.keys())


class ProcessorFactory:
    """Factory for creating appropriate processor instances."""

    _processors = {
        ProcessorType.TEXT: TextProcessor,
        ProcessorType.JSON: JSONProcessor,
    }

    @classmethod
    def create_processor(cls, proc_type: ProcessorType, config: Configuration) -> DataProcessor:
        """Create a processor of specified type."""
        processor_class = cls._processors.get(proc_type)
        if not processor_class:
            raise ValueError(f"Unknown processor type: {proc_type}")
        return processor_class(config)


class ApplicationManager:
    """Main application manager."""

    def __init__(self, config: Configuration):
        """Initialize the application manager."""
        self.config = config
        self.processors: Dict[str, DataProcessor] = {}
        self._initialize_default_processors()

    def _initialize_default_processors(self) -> None:
        """Set up default processors."""
        self.processors["text"] = ProcessorFactory.create_processor(
            ProcessorType.TEXT, self.config
        )
        self.processors["json"] = ProcessorFactory.create_processor(
            ProcessorType.JSON, self.config
        )

    def register_processor(self, name: str, processor: DataProcessor) -> None:
        """Register a custom processor."""
        self.processors[name] = processor

    def process_file(self, filename: str, processor_name: str) -> Optional[str]:
        """Process a file with specified processor."""
        if processor_name not in self.processors:
            raise KeyError(f"Unknown processor: {processor_name}")
        try:
            with open(filename, 'r') as f:
                data = f.read()
            return self.processors[processor_name].process(data)
        except Exception as e:
            if self.config.debug:
                print(f"Error processing {filename}: {e}")
            return None

    def get_processor(self, name: str) -> Optional[DataProcessor]:
        """Retrieve a processor by name."""
        return self.processors.get(name)


def encode_string(text: str) -> str:
    """Encode string to hexadecimal."""
    return ''.join(f'{ord(c):x}' for c in text)


def decode_string(encoded: str) -> str:
    """Decode hexadecimal string."""
    return ''.join(chr(int(encoded[i:i+2], 16)) for i in range(0, len(encoded), 2))


async def process_stream(stream_handler: Any) -> Dict[str, Any]:
    """Process data from an async stream."""
    results = {"processed": 0, "errors": 0}
    async for item in stream_handler:
        try:
            # Process item
            results["processed"] += 1
        except Exception:
            results["errors"] += 1
    return results


if __name__ == "__main__":
    config = Configuration(name="demo", version="1.0", debug=True)
    manager = ApplicationManager(config)
    print("Application initialized successfully")
